challenge-3/src/utils.ts

```diff
@@ -0,0 +1,85 @@
+import { Interface } from "@ethersproject/abi";
+import { BigNumber, Contract, providers } from "ethers";
+import { Alert, AlertQueryOptions, AlertsResponse, Finding, GetAlerts } from "forta-agent";
+import { CHAIN_IDS } from "./constants";
+import * as fs from "fs";
+
+export const getEscrowBalances = async (
+  iface: Interface,
+  provider: providers.Provider,
+  blockNumber: number,
+  l1DaiAddress: string,
+  escrowArbitrumAddress: string,
+  escrowOptimismAddress: string
+): Promise<any> => {
+  const daiContract = new Contract(l1DaiAddress, iface, provider);
+
+  const balanceArbitrum: BigNumber = await daiContract.balanceOf(escrowArbitrumAddress, { blockTag: blockNumber });
+  const balanceOptimism: BigNumber = await daiContract.balanceOf(escrowOptimismAddress, { blockTag: blockNumber });
+
+  return { balanceArbitrum: balanceArbitrum, balanceOptimism: balanceOptimism };
+};
+
+export const verifyInvariant = async (
+  l2DaiAddress: string,
+  chainId: number,
+  alert: Alert,
+  iface: Interface,
+  provider: providers.Provider,
+  blockNumber: number
+): Promise<any> => {
+  const escrowBalance: BigNumber =
+    chainId == CHAIN_IDS.Arbitrum ? alert.metadata.balanceArbitrum : alert.metadata.balanceOptimism;
+  const daiContract = new Contract(l2DaiAddress, iface, provider);
+  const l2DaiSupply: BigNumber = await daiContract.totalSupply({ blockTag: blockNumber });
+
+  //returns if invariant is violated or not
+  return {
+    isViolated: l2DaiSupply.gt(escrowBalance),
+    escrowBalance: escrowBalance,
+    l2DaiSupply: l2DaiSupply,
+  };
+};
+
+//store alerts locally
+export const storeAlerts = async (findings: Finding[], botId: string) => {
+  let alerts: any[] = [];
+  findings.forEach((finding) => {
+    let alert: any = Object.assign({}, finding);
+    alert["botId"] = botId;
+    alerts.push(alert);
+  });
+
+  //write to json file (can be improved)
+  //currently writing only last alert.
+  fs.writeFileSync("alerts.json", JSON.stringify(alerts), "utf-8");
```

Comment:
Thank you for your comments. That's really insightful. Indeed I had the same question of where the finding goes, but I thought using json would mimic calling the forta api to fetch alerts, but that assumes a centralized way of communicating between instances. So since the bot is not live, I need to approach it without testing with real network data  ? (only by mocking getAlerts()) 

Please let me know if I didn't frame my question well.

Reply:
Yes exactly! You could store L1 findings in your tests then access them in your mock of getAlerts(). To limit the scope of your mock to the network layer, I'd recommend maintaining the same type signature as the SDK's getAlerts() function.

Reply:
Alright , thank you !
that was very helpful.